### ISSUE TITLE
Do file.isEmpty and file.nonEmpty have a resource leak?

### ISSUE THREAD

**Issue by rvanheest** [pathikrit/better-files#241]
**Introduction**
I was looking at the source code of this repository. I think it's very clever how you have hidden resource management for the `jStream`s that result from functions such as `Files.list`. As I understand it, a `jStream` is implicitly converted to a Scala `Iterator` like this:

https://github.com/pathikrit/better-files/blob/021384b3d8244aaf57256dd32780b483ed3070af/core/src/main/scala/better/files/Implicits.scala#L340-L342

The `toAutoClosedIterator` that is used here is used in that same file, just a couple of lines up:

https://github.com/pathikrit/better-files/blob/021384b3d8244aaf57256dd32780b483ed3070af/core/src/main/scala/better/files/Implicits.scala#L311-L321

Here is what triggered me in `toAutoClosedIterator`:

> It will NOT close the stream if it is not depleted!

In other words: if you don't iterate through all elements of the `Iterator`, the resource is not being closed.

**Code analysis**
For my use case I wanted to check if a directory is not empty. So the easiest thing to do is to use the `file.nonEmpty` operator. 

https://github.com/pathikrit/better-files/blob/021384b3d8244aaf57256dd32780b483ed3070af/core/src/main/scala/better/files/File.scala#L1079-L1087

When I trace this call to `file.isEmpty`, I find this implementation:

https://github.com/pathikrit/better-files/blob/021384b3d8244aaf57256dd32780b483ed3070af/core/src/main/scala/better/files/File.scala#L1065-L1077

For a directory this calls `file.children.isEmpty`, where `isEmpty` for an `Iterator` is implemented using `!hasNext`. Note that `file.children` implicitly uses the `toAutoClosedIterator` with its 'NOT closed if not depleted' statement.

**Question**
This leads to the following question: do operators like `file.isEmpty` and `file.nonEmpty` close the underlying `jStream` that is produced in `file.children`? After all, they do not deplete the stream with a single `hasNext` call. Therefore: _do operators like `file.isEmpty` and `file.nonEmpty` cause resource leaks?_

**Comment by rvanheest** [comment]
I further looked into this by means of experiments:

1. in IntelliJ I put a breakpoint on the `_.close` lambda in https://github.com/pathikrit/better-files/blob/021384b3d8244aaf57256dd32780b483ed3070af/core/src/main/scala/better/files/ManagedResource.scala#L32.
2. using the code below, I ran the debugger.

```scala
val file = // some path to a directory
println(file.exists) // should print 'true'

// case 1
file.children.toList

// case 2
file.children.take(1).toList // note: the directory has more than 1 child

// case 3
file.nonEmpty
```

For case 1 I found that the breakpoint on `_.close` was hit. However, neither in case 2 nor in case 3  the breakpoint was hit. This is (most probably) due to the underlying `Iterator` not being fully depleted.

I also put similar breakpoints at other lines that I expected the 'dispose process' to hit and none of them where ever hit in cases 2 and 3, while being hit in case 1.
